File: ocaml/mlsearch/tests/searchresulttest.ml
open Core.Std
open OUnit

let test_fixture = "Searchresult" >:::
[
  "singleline_result" >:: (fun () ->
    let pattern = "Search" in
    let file = Searchfile.create "./xsearch/csharp/CsSearch/CsSearch/Searcher.cs" Filetypes.Text in
    let line_num = 10 in
    let matchstartindex = 15 in
    let matchendindex = 23 in
    let line = "\tpublic class Searcher\n" in
    let lines_before = [] in
    let lines_after = [] in
    let result = Searchresult.create pattern line_num matchstartindex matchendindex line lines_before lines_after in
    let result_with_file = { result with file=file } in
    let expected_output = sprintf "%s: %d: [%d:%d]: %s" (Searchfile.to_string file) line_num matchstartindex matchendindex (String.strip line) in
    let resultoutput = Searchresult.to_string result_with_file in
    printf "expected_output: \"%s\"\n" expected_output;
    printf "resultoutput: \"%s\"\n" resultoutput;
    assert_equal expected_output resultoutput
  );

  "multiline_result" >:: (fun () ->
    let pattern = "Search" in
    let file = Searchfile.create "./xsearch/csharp/CsSearch/CsSearch/Searcher.cs" Filetypes.Text in
    let line_num = 10 in
    let matchstartindex = 15 in
    let matchendindex = 23 in
    let line = "\tpublic class Searcher" in
    let lines_before = ["namespace CsSearch"; "{"] in
    let lines_after = ["\t{"; "\t\tprivate readonly FileTypes _fileTypes;"] in
    let result = Searchresult.create pattern line_num matchstartindex matchendindex line lines_before lines_after in
    let result_with_file = { result with file=file } in
    let expected_output = String.concat [
      "================================================================================\n";
      sprintf "%s: %d: [%d:%d]\n" (Searchfile.to_string file) line_num matchstartindex matchendindex;
      "--------------------------------------------------------------------------------\n";
      "   8 | namespace CsSearch\n";
      "   9 | {\n";
      "> 10 | \tpublic class Searcher\n";
      "  11 | \t{\n";
      "  12 | \t\tprivate readonly FileTypes _fileTypes;\n";
    ] in
    let resultoutput = Searchresult.to_string result_with_file in
    printf "expected_output:\n\"%s\"\n" expected_output;
    printf "resultoutput:\n\"%s\"\n" resultoutput;
    assert_equal expected_output resultoutput
  );
]

let _ = run_test_tt ~verbose:true test_fixture;;

File: javascript/jssearch/tests/searchresult_test.js
/*
 * searchresult_test.js
 *
 * Some nodeunit tests of searchresult.js
 */

const common = require('../src/common.js');
const config = require('../src/config.js');
const SearchResult = require('../src/searchresult.js').SearchResult;

const cssearch_path = config.XSEARCHPATH + '/csharp/CsSearch/CsSearch';

exports.testSingleLineResult = function(test) {
    const pattern = 'Search';
    const file = cssearch_path + '/Searcher.cs';
    const linenum = 10;
    const matchStartIndex = 15;
    const matchEndIndex = 23;
    const line = "\tpublic class Searcher\n";
    const linesBefore = [];
    const linesAfter = [];
    const result = new SearchResult(pattern, file, linenum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    const resultString = result.toString();
    const expectedOutput = file + ": " + linenum + ": [" + matchStartIndex + ":" +
        matchEndIndex + "]: " + line.trim();
    test.ok(resultString === expectedOutput, "single-line result matches expected");
    test.done();
};

exports.testBinaryFileResult = function(test) {
    const pattern = 'Search';
    const file = cssearch_path + '/bin/Debug/CsSearch.exe';
    const linenum = 0;
    const matchStartIndex = 5;
    const matchEndIndex = 10;
    const line = "";
    const linesBefore = [];
    const linesAfter = [];
    const result = new SearchResult(pattern, file, linenum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    const resultString = result.toString();
    const expectedOutput = file + " matches at [5:10]";
    test.ok(resultString === expectedOutput, "single-line result matches expected");
    test.done();
};

exports.testMultiLineResult = function(test) {
    const pattern = 'Search';
    const file = cssearch_path + '/Searcher.cs';
    const linenum = 10;
    const matchStartIndex = 15;
    const matchEndIndex = 23;
    const line = "\tpublic class Searcher\n";
    const linesBefore = ["namespace CsSearch\n", "{\n"];
    const linesAfter = ["\t{\n", "\t\tprivate readonly FileTypes _fileTypes;\n"];

    const result = new SearchResult(pattern, file, linenum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    const resultString = result.toString();
    const expectedOutput = "" +
        "================================================================================\n" +
        `${file}: ${linenum}: [${matchStartIndex}:${matchEndIndex}]\n` +
        "--------------------------------------------------------------------------------\n" +
        "   8 | namespace CsSearch\n" +
        "   9 | {\n" +
        "> 10 | \tpublic class Searcher\n" +
        "  11 | \t{\n" +
        "  12 | \t\tprivate readonly FileTypes _fileTypes;\n";
    common.log("\nexpectedOutput:\n\"" + expectedOutput + "\"");
    common.log("\nresultString:\n\"" + resultString + "\"");
    test.ok(resultString === expectedOutput, "multi-line result matches expected");
    test.done();
};
